### PR TITLE
Major import performance improvements

### PR DESCRIPTION
**Pull request by axelstudios**
#### Any background context you want to provide?
Importing data into a large organization can take a considerable amount of time to go through the matching/merging/linking process

#### What's this PR do?
Optimizes several steps throughout the import process to improve performance:

### Step `Matching Data (3/6): Merging Unmatched States`

When importing 512 records that will merge into an org with 512 records, _with UBID data_:

|             | **Before** | **After** |
|-------------|------------|-----------|
| **Queries** | 520        | 7         |
| **Time**    | 1m 28s     | 11s       |

When importing 5000 records that will merge into an org with 145,000 records, _without UBID data_:

|          | **Before** | **After** |
|----------|------------|-----------|
| **Time** | 26m 43s    | 15s     |

When importing 40,000 records that will merge into an org with 145,000 records, _without UBID data_:

### 💥💥💥💥💥

|          | **Before** | **After** |
|----------|------------|-----------|
| **Time** | ~4h 30m    | 30s     |

### Step `Matching Data (4/6): Merging State Pairs`

When importing 512 records that will merge into an org with 512 records, _with UBID data_:

|             | **Before** | **After** |
|-------------|------------|-----------|
| **Queries** | 24,578     | 23,043    |
| **Time**    | 50s        | 22s       |

When importing 5000 records that will merge into an org with 145,000 records, _without UBID data_:

|          | **Before** | **After** |
|----------|------------|-----------|
| **Time** | 2m 35s     | 2m 10s    |

## Total timing when importing 512 records that will merge into an org with 512 records, _with UBID data_:
|             | **Before** | **After** | **With Hannah's improvements** | **With Ross's improvements** |
|-------------|------------|-----------|-----------|-----------|
| **Queries** | 36,222     | 33,600 | 32,571    | 26,913 |
| **Time**    | 1m 24s     | 1m | 57s       | 48s |

#### How should this be manually tested?
Upload files into existing organizations using develop and this branch - compare the runtime

#### What are the relevant tickets?
#4549

### CODE REVIEW

File: seed/data_importer/match.py [hunk 1]
```diff
@@ -548,86 +559,91 @@ def states_to_views(unmatched_state_ids, org, access_level_instance, cycle, Stat
...
+    unmatched_states = StateClass.objects.filter(pk__in=unmatched_state_ids.difference(handled_state_ids))
 
     # For the remaining -States, search for a match within the -States that are attached to -Views.
     # If one match is found, pass that along.
     # If multiple matches are found, merge them together, pass along the resulting record.
-    # Otherwise, add current -State to be promoted as is.
+    # Otherwise, add current -State to be promoted as-is.
     merged_between_existing_count = 0
-    merge_state_pairs = []
+    merge_state_pairs: list[Union[tuple[PropertyState, PropertyState], Union[TaxLotState, TaxLotState]]] = []
```
Comment (haneslinger) [inline seed/data_importer/match.py]: I think that second Union should be a Tuple?

File: seed/data_importer/match.py [hunk 2]
```diff
@@ -548,86 +559,91 @@ def states_to_views(unmatched_state_ids, org, access_level_instance, cycle, Stat
...
-            **matching_criteria,
-        )
+        existing_state_matches = match_lookup.get(tuple(getattr(state, c) for c in tuple_values), [])
 
         if check_jaccard:
             existing_state_matches = [
-                state for state in existing_state_matches if state.ubid and check_jaccard_match(ubid, state.ubid, org.ubid_threshold)
+                state
+                for state in existing_state_matches
+                if state.get("ubid") and check_jaccard_match(matching_criteria.get("ubid"), state.get("ubid"), org.ubid_threshold)
```
Comment (haneslinger) [inline seed/data_importer/match.py:624]: idk how this state and the one declared on L612, arent clashing, but could change this variable name please?

**Review by haneslinger — COMMENTED** [review]
looking good

File: seed/data_importer/match.py (hunk 1 above)
Comment (axelstudios) [reply seed/data_importer/match.py]: Fixed

File: seed/data_importer/match.py (hunk 2 above)
Comment (axelstudios) [reply seed/data_importer/match.py:624]: I'll change it, but variables created within a list comprehension are scoped only within the comprehension itself, so the original `state` variable is unaffected

**Review by haneslinger — APPROVED** [review]

**Pull request merged**